Import the libraries math (for square root and absolute value) and pandas and numpy for reading csv files.

In [12]:
import math
import pandas as pd
import numpy as np

Make an untagged vector, point, and two tagged vectors, data1 and data2:

In [13]:
point = [1, 0, 0, '?'] #(an unknown tag)
data1 = [1, 1, 1, 'M']
data2 = [1, 3, 0, 'F']


Write code to separate the data (X) from the tag (Y).  Your output should be:

The vector [1, 1, 1] has tag M

The vector [1, 3, 0] has tag F


In [14]:
print("The vector " , data1[0:-1], " has tag ", data1[-1])

The vector  [1, 1, 1]  has tag  M


Now let's classify the point as either M or F.  We'll do this by setting k = 1 and using the Euclidean distance.  We'll define that as: 

In [15]:
def euclideanDistance(instance1, instance2, length):
   distance = 0
   for x in range(length):
         #print ('x is ' , x)
         num1=float(instance1[x])
         num2=float(instance2[x])
         distance = distance + pow(num1-num2, 2)
   return math.sqrt(distance)


Now, find out the distance between data1 and point and data2 and point. Output what the tag should be using euclideanDistance.

In [16]:
print(euclideanDistance(data1,point,3))

1.4142135623730951


Now, let's get more data from a file, myFile.csv. This code will read it for us:

In [17]:
url = 'https://github.com/rosenfa/ai/blob/master/myFile.csv?raw=true'
df=pd.read_csv(url,  header=0, error_bad_lines=False) 
#put data in dataset without header line
dataset = np.array(df)

Show that you understand what you read by:

1. Printing the first two vectors in the file
2. Printing the Euclidean distance between those two vectors

Here is some code which I think might help get you on your way!

In [18]:
print(len(dataset))
print(euclideanDistance(dataset[0],dataset[1],3))

10
5.744562646538029


Now assume you have a new value for point:

[0,0,100]

How would you classify this vector using the Euclidean distance function given all of the vectors in the file?

In order to help here is a hint:  

We suggest defining some type of data struction to store different vectors' distances and their tags like this:

In [19]:
class distClass:
    dist = -1 #distance of current point from test point
    tag = '-' #tag of current point


You can then add vector distances like this:

In [20]:
eucDistances = [] # list of distances, will hold objects of type distClass

temp=dataset[1]
label=temp[-1]
d=euclideanDistance(point,temp,3)
print("The distances between " , point , " and " , temp,  " is " , str(d))
print(" and the label is " + label)
obj = distClass() #one record's distance and tag
obj.dist=d
obj.tag=label
eucDistances.append(obj)


The distances between  [1, 0, 0, '?']  and  [1 5 6 'F']  is  7.810249675906654
 and the label is F


In [21]:
eucDistances[0].dist

7.810249675906654

and sort them like this:

In [22]:
eucDistances.sort(key=lambda x: x.dist) 

Questions:

1. What is the label for point if k=1?
2. What is the label for point if k=3?
3. Would the result be different if we used a different distance function like Hamming or Manhattan?

In [23]:
#Add code with functions that implement Hamming and Manhattan distances and test 
#what the label will be for k=1 and k=3 for all possibilities 
#(6 total: 2 Euclidean, 2 Hamming and 2 Manhattan)

Up until here is a simplified version of the homework.
Below here is the work for the part we will be checking as the basis of your grade: 

Now let's look at some bigger files: 

1. mytrain.csv (for training the model)
2. mytest.csv (for testing)


In [24]:
url = 'https://github.com/rosenfa/ai/blob/master/mytrain.csv?raw=true'
train_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))
url = 'https://github.com/rosenfa/ai/blob/master/mytest.csv?raw=true'
test_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))

In [25]:
print(train_data.shape)# number of records and features
print(train_data)

(100, 30)
[[0.317251 0.540016 0.150514 ... 0.0 0.212007 'F']
 [3.882181 0.0 0.0 ... 0.0 0.0 'F']
 [1.790708 0.0 0.0 ... 0.0 0.0 'F']
 ...
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'F']]


In [26]:
print(test_data.shape)# number of records and features
print(test_data)

(100, 30)
[[2.85985 0.66275 0.0 ... 0.995289 0.0 'F']
 [0.0 12.998726 0.0 ... 0.0 0.0 'F']
 [13.231318 0.0 0.0 ... 0.0 0.0 'F']
 ...
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [1.688611 0.0 3.204529 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'M']]


I hope by now you understand where we are going with this :)

Now implement the knn code with 3 different values for k:
1. k = 1
2. k = 7
3. k = 15

and at first use the Euclidean distance.
Classify each of the vectors in the test_data dataset using the training data from train_data.  Which value for k did the best?  What accuracy did it give you?

Now see if using Hamming or Manhattan distance give any better results for the same values of k.  

Once you are done, you should have a total of 9 different results:

1. Three results for the different value of k using the Euclidean Distance
2. Three results for the different value of k using the Hamming Distance
3. Three results for the different value of k using the Manhattan Distance

Hint: I strongly suggest you base yourself on the code you've seen until this point so you don't have to reinvent the wheel!

Defining the hamming and manhattan distance:

In [27]:
def hammingDistance(instance1, instance2, length):
  count = 0; 
  for x in range(length):
    num1=float(instance1[x])
    num2=float(instance2[x])
    count += int(num1 != num2)
  return count

def manhattanDistance(instance1, instance2, length):
  count = 0; 
  for x in range(length):
    num1=float(instance1[x])
    num2=float(instance2[x])
    count += abs(num1 - num2)
  return count

define the method knn:

In [28]:
def knn(distances, k):
  distances.sort(key=lambda x: x.dist) 
  M = 0
  F = 0
  for i in range(k):
    if distances[i].tag == 'M' :
      M+=1
    else:
      F+=1
  if M > F :
    return 'M'
  else:
    return 'F'

In [29]:
correct = 0
not_correct = 0
for point_test in test_data:
  distances = []
  for point_train in train_data:
    obj = distClass()
    obj.dist = manhattanDistance(point_test, point_train, 29) # change distance function, either euclidean, hamming, or manhattan
    obj.tag = point_train[29]
    distances.append(obj)
  if knn(distances, 7) == point_test[29]: # change k for different tests, either 1, 7, or 15
    correct+=1

print(correct/100)

0.63


# **Results:**

Euclidian:

k = 1 - 50% accuracy 

k = 7 - 74% accuracy

k = 15 - 70% accuracy


Hamming:

k = 1 - 61% accuracy 

k = 7 - 55% accuracy

k = 15 - 57% accuracy

Manhattan:

k = 1 - 61% accuracy 

k = 7 - 63% accuracy

k = 15 - 69% accuracy

Grade Basis:

80% for correct answers (and yes, there are possibilities that multiple answers are correct-- especially for cases of ties).

20% : Documentation and easily readable code

Please publish your final Notebook in your Github directory.

The homework assignment is due by November 30th.
